File: solutions/elixir/158-random-pick-with-weight.ex
# Problem 158: Random Pick with Weight (LeetCode 528)
# Difficulty: Med
# Language: Elixir
# 
defmodule WeightedPicker do
  def build(weights) do
    weights |> Enum.scan(&+/2)
  end
  def pick(prefix) do
    total = List.last(prefix)
    target = :rand.uniform(total)
    binary_search(prefix, target, 0, length(prefix) - 1)
  end
  defp binary_search(ps, target, lo, hi) when lo >= hi, do: lo
  defp binary_search(ps, target, lo, hi) do
    mid = div(lo + hi, 2)
    if Enum.at(ps, mid) < target, do: binary_search(ps, target, mid + 1, hi),
    else: binary_search(ps, target, lo, mid)
  end
end

File: solutions/scala/158-random-pick-with-weight.scala
/**
 * Problem 158: Random Pick with Weight (LeetCode 528)
 * Difficulty: Med
 * Language: Scala
 */
class WeightedPicker(w: Array[Int]) {
  private val prefix = w.scanLeft(0)(_ + _).tail
  def pickIndex(): Int = {
    val target = scala.util.Random.nextInt(prefix.last) + 1
    def search(lo: Int, hi: Int): Int =
      if (lo >= hi) lo
      else { val mid = (lo + hi) / 2
        if (prefix(mid) < target) search(mid + 1, hi) else search(lo, mid) }
    search(0, prefix.length - 1)
  }
}
